cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -263,6 +270,389 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
+{
+public:
+    MinkowskiDistances(const NumericTable & a, const NumericTable & b, const bool powered = true, const double p = 2.0)
+        : _a(a), _b(b), _powered(powered), _p(p)
+    {}
+
+    virtual ~MinkowskiDistances() {}
+
+    virtual services::Status init()
+    {
+        services::Status s;
+        return s;
+    }
+
+    virtual services::Status computeBatch(const FPType * const a, const FPType * const b, size_t aOffset, size_t aSize, size_t bOffset, size_t bSize,
+                                          FPType * const res)
+    {
+        computeBatchImpl(a, b, aOffset, aSize, bOffset, bSize, res);
+
+        return services::Status();
+    }
+
+protected:
+    FPType computeDistance(const FPType * x, const FPType * y, const size_t n)
+    {
+        daal::internal::mkl::MklMath<FPType, cpu> math;
+
+        FPType d = 0;
+
+        if (_p == 1.0)
+        {
+            for (size_t i = 0; i < n; ++i)
+            {
+                d += math.sFabs(x[i] - y[i]);
+            }
+
+            return d;
+        }
+        else
+        {
+            for (size_t i = 0; i < n; ++i)
+            {
+                d += math.sPowx(math.sFabs(x[i] - y[i]), _p);
+            }
+
+            if (!_powered) return math.sPowx(d, 1.0 / _p);
+
+            return d;
+        }
+    }
+
+    services::Status computeBatchImpl(const FPType * const a, const FPType * const b, size_t aOffset, size_t aSize, size_t bOffset, size_t bSize,
+                                      FPType * const res)
+    {
+        daal::internal::mkl::MklMath<FPType, cpu> math;
+
+        const size_t nDims = _a.getNumberOfColumns();
+        const size_t nX    = aSize;
+        const size_t nY    = bSize;
+
+        const FPType * x = a;
+        DAAL_CHECK_MALLOC(x);
+
+        const FPType * y = b;
+        DAAL_CHECK_MALLOC(y);
+
+        const size_t BlockSize        = 64;
+        const size_t THREADING_BORDER = 32768;
+        const size_t nBlocksX         = nX / BlockSize;
+        const size_t nBlocksY         = nY / BlockSize;
+        const size_t nThreads         = threader_get_threads_number();
+
+        if (nThreads > 1 && nX * nY > THREADING_BORDER)
+        {
+            daal::threader_for(nBlocksX, nBlocksX, [&](size_t iBlockX) {
+                const size_t startX = iBlockX * BlockSize;
+                const size_t endX   = (nBlocksX - iBlockX - 1) ? startX + BlockSize : nX;
+
+                daal::threader_for(nBlocksY, nBlocksY, [&](size_t iBlockY) {
+                    const size_t startY = iBlockY * BlockSize;
+                    const size_t endY   = (nBlocksY - iBlockY - 1) ? startY + BlockSize : nY;
+
+                    for (size_t ix = startX; ix < endX; ++ix)
+                        for (size_t iy = startY; iy < endY; ++iy)
+                        {
+                            res[ix * nY + iy] = computeDistance(x + ix * nDims, y + iy * nDims, nDims);
+                        }
```

Comment:
```suggestion
                    for (size_t ix = startX; ix < endX; ++ix)
                    {
                        for (size_t iy = startY; iy < endY; ++iy)
                        {
                            res[ix * nY + iy] = computeDistance(x + ix * nDims, y + iy * nDims, nDims);
                        }
                    }
```

Reply:
Moreover what about fetching this vector?

Reply:
done for all new "for" across this file